File: leetcode/JavaScript/maximumDepthOfBinaryTree.js
class Node {
  constructor(val) {
    this.val = val
    this.left = null
    this.right = null
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null
  }
  insert(val) {
    const newNode = new Node(val)
    if (!this.root) {
      this.root = newNode
      return this
    }
    try {
      let currentNode = this.root
      while (true) {
        if (val === currentNode.val) throw new Error("Node already present")
        if (val < currentNode.val) {
          if (!currentNode.left) {
            currentNode.left = newNode
            return this
          }
          currentNode = currentNode.left
        } else if (val > currentNode.val) {
          if (!currentNode.right) {
            currentNode.right = newNode
            return this
          }
          currentNode = currentNode.right
        }
      }
    } catch (e) {
      console.log(e.message)
    }
  }
  contains(val) {
    if (!this.root) return false
    let currentNode = this.root
    let found = false
    while (currentNode && !found) {
      if (val < currentNode.val) {
        currentNode = currentNode.left
      } else if (val > currentNode.val) {
        currentNode = currentNode.right
      } else {
        found = true
      }
    }
    return found
  }
}

const tree = new BinarySearchTree()
tree.insert(5)
tree.insert(4)
tree.insert(3)
tree.insert(2)
tree.insert(1)

// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  if (!root) return 0
  if (!root.left && !root.right) return 1

  let max = 0

  let queue = [root]

  while (queue.length > 0) {
    const subQueue = queue
    queue = []
    while (subQueue.length > 0) {
      const currentNode = subQueue.shift()
      if (currentNode.left) queue.push(currentNode.left)
      if (currentNode.right) queue.push(currentNode.right)
    }
    max++
  }

  return max
}
// @leet end

console.log(maxDepth(tree.root))

File: leetcode/python/maximumDepthOfBinaryTree.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def recursiveMaxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        return 1 + max(
            self.recursiveMaxDepth(root.left), self.recursiveMaxDepth(root.right)
        )

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        if not root.right and not root.left:
            return 1

        queue: List[TreeNode] = []
        queue.append(root)
        max = 0

        while len(queue) > 0:
            subQueue: List[TreeNode] = queue
            queue = []

            while len(subQueue) > 0:
                node = subQueue.pop()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            max += 1

        return max


root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
root = TreeNode(1, None, TreeNode(2))
root = TreeNode(1, TreeNode(2, TreeNode(4)), TreeNode(3, None, TreeNode(5)))
solution = Solution()
print(solution.recursiveMaxDepth(root))

"""
This implementation uses BFS to calculate the max depth
We create a queue and we push the root inside.
Then we do a nested loop:
We copy the current queue and reset the main one.
While the copy is more than 0:
We pop from the copy and push left and right of the popped node (if they exist) to the main queue.
After the inner loop, we update max because we're sure we processed a level of the tree.
We keep going until we traversed all the tree.
We return the max variable.

The important part in this implementation is how we know that we processed an entire level before adding one to max.
We know thid because at every cycle of the outer while loop, we create a copy of the current main queue. That copy contains all the nodes of the current level. This means that when we process all the copy queue, we know for sure that we processed all the level and populated the next one in the main queue.

The time complexity is O(n) because we process every node only once
The space complexity is O(w) where n is the max width of the tree because we always keep in memory a level at a time
"""
